stablehlo/reference/Ops.cpp

```diff
@@ -33,13 +34,71 @@ namespace mlir {
 namespace stablehlo {
 namespace {
 
-Index evalIndices(ArrayRef<Tensor> runtimeIndices) {
-  Index index(runtimeIndices.size());
-  for (size_t i = 0; i < runtimeIndices.size(); ++i)
-    index[i] = runtimeIndices[i].get({}).getIntegerValue().getSExtValue();
+Index evalIndices(ArrayRef<Tensor> indices) {
+  Index index(indices.size());
+  for (size_t i = 0; i < indices.size(); ++i)
+    index[i] = indices[i].get({}).getIntegerValue().getSExtValue();
   return index;
 }
 
+Tensor evalPadOp(const Tensor &operand, const Tensor &paddingValue,
+                 const Sizes &edgePaddingLow, const Sizes &edgePaddingHigh,
+                 const Sizes &interiorPadding) {
+  SmallVector<Type> inferredTypes;
+  Builder builder(operand.getType().getContext());
+  auto inferStatus =
+      hlo::inferPadOp({}, operand.getType(), paddingValue.getType(),
+                      builder.getI64TensorAttr(edgePaddingLow),
+                      builder.getI64TensorAttr(edgePaddingHigh),
+                      builder.getI64TensorAttr(interiorPadding), inferredTypes);
+  if (failed(inferStatus))
+    report_fatal_error(invalidArgument("Could not infer PadOp's return type"));
+  return evalPadOp(operand, paddingValue, edgePaddingLow, interiorPadding,
+                   inferredTypes[0].cast<ShapedType>());
+}
+
+SmallVector<Tensor> evalReduceOp(ArrayRef<Tensor> inputs,
+                                 ArrayRef<Tensor> initValues,
+                                 const Axes &dimensions, Region &body,
+                                 Scope &scope) {
+  SmallVector<Type> inputTypes;
+  for (auto input : inputs) inputTypes.push_back(input.getType());
+
+  SmallVector<Type> initValueTypes;
+  for (auto initValue : initValues)
+    initValueTypes.push_back(initValue.getType());
+
+  SmallVector<ShapedTypeComponents> inferredReduceTypes;
+  Builder builder(inputs[0].getType().getContext());
+  auto reduceStatus = hlo::inferReduceOp(
+      /*location=*/{}, inputTypes, initValueTypes,
+      builder.getI64TensorAttr(dimensions), inferredReduceTypes);
+  if (failed(reduceStatus))
+    report_fatal_error(
+        invalidArgument("Could not infer ReduceOp's return type"));
+
+  SmallVector<ShapedType> resultTypes;
+  for (auto inferredType : inferredReduceTypes)
+    resultTypes.push_back(RankedTensorType::get(inferredType.getDims(),
+                                                inferredType.getElementType()));
```

Comment:
Now that I'm looking at this again, I'm realizing that this logic is incomplete, e.g. it doesn't account for encoding, and it doesn't account for missing pieces.

How about we add a function to Base.h, next to unflattenTupleTypes, with the following signature `ShapedType createShapedType(ShapedTypeComponents components)` and the following behavior:
  * If elementType is null, then return null.
  * If hasRank, then created a RankedTensorType from dims, elementType and attribute.
  * If !hasRank, then created an UnrankedTensorType from elementType.

Then here we would call this function, check if it returned null (and report_fatal_error if that's the case) and if not, push back its result.

Reply:
Ended up having a slightly different signature. PTAL.